### ISSUE TITLE
Fix the test that checks if core projects have `type =` specified in their .info files

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#3829]
This is a follow-up to #3793/#3821 

@jenlampton:
> @quicksketch we should check our test too (the one that checks that all modules `.info` files have a type line) and make sure it checks test modules too.

Although we have fixed the immediate problem in #3793, we need to look into why this was not picked up by the test, and fix it, so that this does not happen again.

---
PR by @klonos: https://github.com/backdrop/backdrop/pull/2707

**Comment by klonos** [comment]
...working on this currently

**Comment by klonos** [comment]
So here's the thing...

- in `testModuleInfoFileContents()`, we do this:
  `$modules = backdrop_system_listing('/^' . BACKDROP_PHP_FUNCTION_PATTERN . '\.module$/', 'core/modules', 'name', 0);`
- in `testThemeInfoFileContents()`, we do this:
  `$themes = backdrop_system_listing('/^' . BACKDROP_PHP_FUNCTION_PATTERN . '\.info$/', 'core/themes');`
- in `testLayoutInfoFileContents()`, we do this:
  `$layouts = backdrop_system_listing('/^' . BACKDROP_PHP_FUNCTION_PATTERN . '\.info$/', 'core/layouts', 'name', 0);`

So we are not testing for anything inside `/core/profiles` at all, and we are making the following assumptions:

- any .info file under `/core/modules` belongs to a module
- any .info file under `/core/themes` belongs to a theme 
- any .info file under `/core/layouts` belongs to a layout

So basically, we do not account for example for:
- themes (or layouts) located within module directories (like `/core/modules/simpletest/tests/themes` for example)
- modules (or themes, or layouts) located within profile directories (`/core/profiles/testing/modules` for example)

We could start tweaking the logic, but it feels like things will get overly complicated really quick. We'd still end up making assumptions, most likely need to make exceptions, and I personally do not see the benefit of doing this.

Wouldn't it be much simpler to just have a single test that tests *all* .info files for `type =`? Why do we test for specific types of projects under those specific directories, when we are in fact not actually enforcing that?

**Comment by klonos** [comment]
PS: in all 3 tests, we test for `!isset($info['type']` but then have all respective instances of `$this->fail(format_string('%key does not have project type set to "xyz".', array('%key' => $key)));` (where xyz is module/theme/layout); whereas we should say something like `%key does not specify a project type` instead.

**Comment by quicksketch** [comment]
Thanks @klonos! Great work expanding that test coverage, and you made it simpler/faster at the same time! Hopefully we've now seen the end of accidentally breaking the self-updater due to this issue. Merged into 1.x and 1.13.x.
